pgx/src/array.rs

```diff
@@ -0,0 +1,235 @@
+    // so it requires validity assertions in order to be constructed.
+    // The main reason it uses NonNull and denies Clone, however, is access soundness:
+    // It is not sound to go from &mut Type to *mut T if *mut T is not a field of Type.
+    // This creates an obvious complication for handing out pointers into varlenas.
+    // Thus also why this does not use lifetime-bounded borrows.
+
+    /// Returns a handle to the raw array header.
+    ///
+    /// # Safety
+    ///
+    /// When calling this method, you have to ensure that all of the following is true:
+    /// * The pointer must be properly aligned.
+    /// * It must be "dereferenceable" in the sense defined in [the std documentation].
+    /// * The pointer must point to an initialized instance of `ArrayType`.
+    /// * This can be considered a unique, "owning pointer",
+    ///   so it won't be aliased while RawArray is held,
+    ///   and it points to data in the Postgres ArrayType format.
+    ///
+    /// It should be noted as RawArray is not inherently lifetime-bound, it can be racy and unsafe!
+    ///
+    /// [the std documentation]: core::ptr#safety
+    pub unsafe fn from_raw(at: NonNull<ArrayType>) -> RawArray {
+        RawArray { at }
+    }
+
+    /// # Safety
+    ///
+    /// Array must have been constructed from an ArrayType pointer.
+    pub unsafe fn from_array<T: FromDatum>(arr: Array<T>) -> Option<RawArray> {
+        let array_type = arr.into_array_type() as *mut _;
+        Some(RawArray {
+            at: NonNull::new(array_type)?,
+        })
+    }
+
+    /// Returns the inner raw pointer to the ArrayType.
+    pub fn into_raw(self) -> NonNull<ArrayType> {
+        self.at
+    }
+
+    /// Get the number of dimensions.
+    pub fn ndims(&self) -> libc::c_int {
+        // SAFETY: Validity asserted on construction.
+        unsafe {
+            (*self.at.as_ptr()).ndim
+            // FIXME: While this is a c_int, the max ndim is normally 6
+            // While the value can be set higher, it is... unlikely
+            // that it is going to actually challenge even 16-bit pointer widths.
+            // It would be preferable to return a usize instead,
+            // however, PGX has trouble with that, unfortunately.
+            as _
+        }
+    }
+
+    /// A raw slice of the dimensions.
+    /// Oxidized form of ARR_DIMS(ArrayType*)
+    ///
+    /// # Safety
+    ///
+    /// This requires &mut to collect the slice, but be aware: this is a raw pointer!
+    /// If you find ways to store it, you are probably violating ownership.
+    /// Raw pointer validity is **asserted on dereference, not construction**,
+    /// so remember, this slice is only guaranteed to be valid almost immediately
+    /// after obtaining it, or if you continue to hold the RawArray.
+    pub fn dims(&mut self) -> NonNull<[libc::c_int]> {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_DIMS
```

Comment:
Should this path be visible in the public documentation? I suppose they can click through to see the source, but the public docs will support links.

Reply:
Yeah, actually, directly linking is a good idea. And that way we could also link to the doxygen.

Reply:
Doxygen is too unstable, will point to master and lines might change. Instead used gitweb links.

Reply:
From a pgx-maintainers perspective, how much of a maintenance nightmare is it linking to the doxygen pages? How often would that change?

Reply:
Potentially every time someone lands a commit into postgresql/master.